Document package types (#2246)
jpmorganchase/modular: docs/commands/add.md

# `modular add <packageName>`

> Adds a new package by creating a new package at the workspace located at

Reviewer: Does it _always_ use `packages` (when --path is _not_ supplied)? I'm thinking about the use case where someone has defined custom yarn `workspaces` values in their `package.json`

Author: Yep, we hardcode the package root to `packages` if no `--path` option is provided. Maybe it can be done better (for example, forcing the user to specify a `--path` if the `workspaces` entry doesn't contain `packages`), but I'd rather think of it when we support many package manager, because that's when it becomes complicated.

> Adds a new package by creating a new package at the workspace located at

Author: "Adds" is a bit unclear, it suggests more than one package could be in a workspace.

`packages/<packageName>`, omitting the scope if the package is
[scoped](https://docs.npmjs.com/cli/v8/using-npm/scope). If `--path <somePath>`
is specified, the command creates the workspace at `<somePath>/<packageName>`.

(i.e. `modular add my-app` would create a package in `packages/my-app`,
`modular add @scoped/my-scoped-app` would create a package in
`packages/my-scoped-app` and `modular add lib-a --path libs` would create a
package in `libs/lib-a`)

The `modular add` command prompts the user to choose the Modular `type` of the
package it's about to create. The next section briefly describes the various
types that can be created by the `modular add` command. For an in-depth
discussion of the available package types and their characteristics, please see
[this page](../concepts/package-types.md).

> ### Standalone (bundled) package types

Reviewer 2: The docs written here covering package types should be moved into the `docs/package-types/index.md` file as I have proposed in another comment. The `add.md` page should link to this, but I think the concepts will be clearer if we separate out the commands from the types rather than mixing them rather thoroughly as we have done.

Author: Types are here to provide quick info to developers; new users will run `modular add` for the first time and, when prompted for package type, will go to the "add" command documentation since that's what they run. I think it makes sense to provide a short summary of each type here. See it as a redundancy feature to quickly give the information needed.

I added links to each of the type pages in the type description; it should be easy for users to follow the link if they want more info.

These package types are built with [Webpack v5](https://webpack.js.org/) or, if
specified in the [configuration](../configuration.md),

> [esbuild](https://esbuild.github.io/). Modules imported in the source these

Reviewer: Possibly a typo?

```suggestion
[esbuild](https://esbuild.github.io/). Modules imported in the source of these
```

package types are bundled in the final result or, in case of `esm-view`s, they

> can be rewritten to an [external ESM CDN](../esm-views/esm-cdn.md).

Reviewer: > can be written to an external ESM CDN

when? is the idea to follow the esm-cdn docs to understand this answer?

Author: Agree it's a little hand-wavy. I rewrote it in a more informative way and liked the esm views section for more clarity.

- `app`. This package type corresponds to a static Single Page Application (SPA)

>   project in a workspace. It's possible to specify a custom `index.html` file

Reviewer: > It's possible to specify a custom index.html file and public assets in the public directory?

When/how? Maybe just a link to other docs needed

Author: I linked this to the `package-types` page, where we essentially say we're compatible with CRA and we list the differences. It's a fine balance between being clear and not re-writing the CRA docs :)

>   and public asstets in the `public` directory.

Reviewer: ```suggestion
  and public assets in the `public` directory.
```

- `esm-view`. This package type is an app that gets built as an ES module that
  can be imported at runtime. `esm-view`s are typically used to implement a
  [micro-frontend](../concepts/microfrontends.md) architecture. `esm-views`,
  when [built](./build.md) or [started](./start.md) will also generate a
  `index.html` file that tries to load the ES Module and render its default
  export as a React component onto the DOM (standalone mode). See also
  [the esm-view reference](../esm-views/index.md) for an in-depth introduction.

### Library package types

These package types are either built with

> [Rollup.js](https://rollupjs.org/guide/en/) as CJS and ES Modules or, in case of

Reviewer: Might be sensible to avoid an acronym for commonjs

```suggestion
[Rollup.js](https://rollupjs.org/guide/en/) as CommonJS and ES Modules or, in case of
```

> `source` modules, they are not built at all. Packages with these types get

Reviewer: ```suggestion
`source` modules, they are not built at all. Packages with the `source` type get
```

Author: Reworded, distinguishing between package/view and source

> normally published to NPM or get imported by other packages in the monorepo. For

Reviewer: Wording here is currently a bit tricky. What about this?

```suggestion
published to NPM (as regular npm packages) or get imported by other packages in the monorepo. For
```

Author: Reworded, as above

> this reason, files are transpiled separately on buld and external dependencies

Reviewer: ```suggestion
this reason, files are transpiled separately on build and external dependencies
```

> are never "pulled in" (no bundling).

Reviewer: Could be a bit more explicit?

```suggestion
are never "pulled in" (i.e. not included in a bundle).
```

- `package`. This is a generic package with a single entry point. It's normally

>   used to create a publishable library that gets transpiled to Common JS and ES

Reviewer: Consistent naming of CommonJS

```suggestion
  used to create a publishable library that gets transpiled to CommonJS and ES
```

  [this page](../package-types/#app) for more information about apps.

- [`esm-view`](../package-types/esm-view.md). This package type is an app that
  gets built as an ES module that can be imported at runtime. `esm-view`s are
  typically used to implement a [micro-frontend](../concepts/microfrontends.md)
  architecture. `esm-views`, when [built](./build.md) or [started](./start.md)
  will also generate a `index.html` file that tries to load the ES Module and
  render its default export as a React component onto the DOM (standalone mode).
  See also [the esm-view reference](../esm-views/index.md) for an in-depth
  introduction.

### Library package types

These package types are either built with
[Rollup.js](https://rollupjs.org/guide/en/) as CommonJS and ES Modules or, in
case of `source` modules, they are not built at all. Library package types get
typically published to NPM (`package` and `view` types) or get imported by other
packages in the monorepo (`source` type). For this reason, files are transpiled
separately on build and external dependencies are never "pulled in" (i.e. not
included in a bundle).

> - [`package`](../package-types/package.md). This is a generic package with a

Author: Since our audience are developers, I hope they can extrapolate the "why" from the "what". If we go the other way we are forced to make examples ("use an app if you want to create a standalone trading application, use a view if you want to make a chart component") that are always imprecise and will be misinterpreted or used to ask for custom, specific functionality. We already purged the docs from this kind of examples (eg "express server" for `package`) and it works a bit better for us. If you think this should be expressed better, feel free to propose a new version of this paragraph!

> - [`package`](../package-types/package.md). This is a generic package with a

Reviewer 2: That's an interesting point; it had not occurred to me that the "why" would be less precise, but I totally see what you mean. I need to let that sit with me for a while, so I'll resolve this conversation.
